### PR TITLE
Add a counting bloom filter package

### PR DESCRIPTION
**Pull request by Tryneus**
This PR adds `src/server/pkg/bloom`, which implements a counting bloom filter.  This filter can be instantiated by specifying one of these sets of constraints:
* element count and memory footprint - false positive rate will be derived from these
* element count and false positive rate - memory footprint will be derived from these

In order to work with the chunks, which are indexed by their hash, this bloom filter does not do hashing itself, but accepts hashes to any member functions.

The main interface for the filter is:
```go
func (f *BloomFilter) Add(hash []byte) {...}
func (f *BloomFilter) Remove(hash []byte) {...}
func (f *BloomFilter) IsNotPresent(hash []byte) bool {...}
func (f *BloomFilter) UpperBoundCount(hash []byte) int {...}
```

This is a counting bloom filter, which means that we can (in theory) sum them together, although this behavior was not implemented as we don't have a use case at the moment.

The size of each bucket was chosen to be 1-byte as that gives us a fairly good runway for reasonable false-positive rates without making the memory footprint too large.  Once a bucket overflows (i.e. hits the maximum value), we will never subtract from that bucket as we can no longer make guarantees about the number of values hitting that bucket).

### CODE REVIEW

File: src/server/pkg/bloom/bloom_test.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+func requirePanic(t *testing.T, cb func()) {
+	defer func() {
+		r := recover()
+		require.NotNil(t, r)
+	}()
+
+	cb()
+}
```
Comment (jdoliner) [inline src/server/pkg/bloom/bloom_test.go]: Might as well just move this into the `require` pkg, we'll probably need it elsewhere eventually.

**Comment by jdoliner** [comment]
I think overflow might be an issue with this design, we expect objects to have a lot of references, because there are going to be pieces of data that get committed once and then never changed which could be referenced by every single commit in a repo, and then the entire repo could get deleted. Maybe this isn't as much of a problem as I think it is, because we can always recompute the bloom filters by merging them. However it might also make more sense to just build more resolution into the system from the start, especially because we can use `varint` encoding (which proto3 does, and go has in its standard library) to decrease the space needed for smaller numbers.

**Comment by Tryneus** [comment]
Oh, interesting - I wasn't aware golang had varint support.  I'll integrate that, bucket overflow is going to be a major concern otherwise.  Do you think it would be a good idea to support negative numbers in buckets (so that we can have a 'delta' bloom filter), or is that overkill?

**Comment by Tryneus** [comment]
Because it'd be difficult to efficiently use `varint`s in-memory, and to avoid an extra serialization step, I bumped the bucket size to 4 bytes (`uint32`).  Proto3 should still efficiently store the data on-disk using `varint`, though.

File: src/client/pkg/require/require.go [hunk 2]
```diff
@@ -413,6 +413,18 @@ func False(tb testing.TB, value bool, msgAndArgs ...interface{}) {
 	}
 }
 
+// Panic checks that the callback panics.
+func Panic(tb testing.TB, cb func(), msgAndArgs ...interface{}) {
```
Comment (jdoliner) [inline src/client/pkg/require/require.go]: Can we call this `YesPanic` to match with `YesError`?

**Pull request merged**